test_plans/buffers.asciidoc

```diff
@@ -0,0 +1,94 @@
+
+In full conformance mode:
+
+* `char`
+* `signed char`
+* `unsigned char`
+* `short int`
+* `unsigned short int`
+* `unsigned int`
+* `long int`
+* `unsigned long int`
+* `long long int`
+* `unsigned long long int`
+* `bool`
+* A user-defined struct with several scalar member variables, no constructor, destructor or member functions.
+* A user-defined class with several scalar member variables, a deleted default constructor, and a user-defined (non-default) constructor.
+
+In addition, if the device has `aspect::fp64`, the following type is tested:
+
+* `double`
+
+In addition, if the device has `aspect::fp16`, the following type is tested:
+
+* `sycl::half`
+
+== Tests
+
+=== SYCL1.2.1 compatible interface
+
+CTS contains tests for `buffer` SYCL1.2.1 interface. Some functions have been marked as deprecated in SYCL2020. These tests should be marked as deprecated and be enrolled into the machinery disabling legacy tests. These functions are the following:
+
+* `size_t get_count()`
+* `size_t get_size()`
+* `template <access_mode Mode> accessor<T, Dimensions, Mode, target::host_buffer> get_access()`
+* `template <access_mode Mode> accessor<T, Dimensions, Mode, target::host_buffer> get_access(range<Dimensions> accessRange, id<Dimensions> accessOffset = {})`
+
+=== SYCL2020 interface
+
+==== Container constructors
+
+For `Dimensions == 1` check that:
+
+* It is possible to provide a non-empty, non-`const T` `std::vector` instance to the `template <typename Container> buffer(Container& container, const property_list& propList = {})` constructor.
+* After modifying its contents on device, updated contents are written back to the vector upon buffer destruction.
+
+For `AllocatorT == sycl_cts::logging_allocator`
+
+* It is possible to provide a non-empty, non-`const T` `std::vector` instance and a custom logging allocator to the `template <typename Container> buffer(Container& container, AllocatorT allocator, const property_list& propList = {})` constructor.
+* After modifying its contents on device, updated contents are written back to the vector upon buffer destruction.
+* The logging allocator's log isn't empty.
+
+=== Linearization
+
+For `Dimensions == {2;3}` check that:
+
+* An appropriately sized multi-dimensional buffer initialized with multi-dim indices when read back to host is linearized according to 3.11.1
+
+=== no_init property
+
+For `Dimensions == 1` check that:
+
+* When fully initializing a buffer on device, a subsequent kernel adding a requirement on the buffer by creating a `no_init` `write_only` accessor really calls the only trivial CTOR for every element, even without explictly filling the buffer.
+* When fully initializing a buffer on device, a subsequent kernel adding a requirement on the buffer by creating a `no_init` `write_only` ranged accessor really calls the only trivial CTOR for the covered range, even without explictly filling the buffer.
```

Comment:
I don't understand these tests.  A typical SYCL implementation does not ever call the `T` constructor when copying data between host and device.  This is the reason we limit `T` to "device copyable" types.  The idea is that an implementation can just move data via direct memcpy.  This behavior is the same regardless of whether the `no_init` property is specified.

Even when `no_init` is specified, it's only a hint to the implementation that *can* enable optimization.  A conformant implementation is allowed to ignore the property if it wants.  The SYCL spec doesn't provide any guarantee that an optimization occurs.

The spec does provide a guarantee related to `no_init` that would be interesting to test:

> If this is a ranged accessor, this applies only to the elements within the accessor’s range. The values of unwritten elements outside of this range are preserved.

Therefore, an interesting test could be:

* Construct a buffer with initialized values.
* Create a *ranged* accessor to that buffer with `no_init`.
* Pass the accessor to a kernel that does not write any values through the accessor.
* Make sure that the elements *outside* of the range still have their initial values.

Reply:
Intel seems to have a working set of no_init tests internally that'll get upstreamed sometime in the future, so it's safe to remove it from this plan.